src/aapns/pool.py

```diff
@@ -0,0 +1,218 @@
+        bits.append(f"errors:{self.errors}")
+        return "<Pool %s>" % " ".join(bits)
+
+    @classmethod
+    async def create(cls, origin: str, size=2, ssl=None):
+        if size < 1:
+            raise ValueError("Connection pool size must be strictly positive")
+        ssl_context = ssl or create_ssl_context()
+        connections = set(
+            await gather(
+                *(Connection.create(origin, ssl=ssl_context) for i in range(size))
+            )
+        )
+        # FIXME run the hook / ensure no connection is dead
+        return cls(origin, size, ssl_context, connections)
+
+    def __post_init__(self):
+        self.maintenance = create_task(self.maintain(), name="maintenance")
+
+    @property
+    def state(self):
+        if not self.closing:
+            return "active"
+        elif not self.closed:
+            return "closing"
+        else:
+            return "closed"
+
+    def resize(self, size):
+        assert size > 0
+        self.size = size
+        self.maintenance_needed.set()
+
+    def cert_hook(self, c):
+        if not self.outcome and c.outcome == "BadCertificateEnvironment":
+            self.closing = True
+            self.outcome = c.outcome
+
+    async def maintain(self):
+        while True:
+            if self.closing or self.closed:
+                return
+
+            for c in list(self.active):
+                if c.closing:
+                    self.active.remove(c)
+                    self.dying.add(c)
+                    self.cert_hook(c)
+
+            while len(self.active) > self.size:
+                c = self.active.pop()
+                c.closing = True
+                self.dying.add(c)
+                self.cert_hook(c)
+
+            for c in list(self.dying):
+                if c.closed:
+                    self.dying.remove(c)
+                    self.cert_hook(c)
+                elif not c.channels:
+                    self.dying.remove(c)
+                    try:
+                        await c.close()
+                    finally:
+                        self.cert_hook(c)
+                if self.closing or self.closed:
+                    return
+
+            while len(self.active) < self.size:
+                if not await self.add_one_connection():
+                    break
+                if self.closing or self.closed:
+                    return
+
+            # FIXME wait for a trigger:
+            # * some connection state has changed
+            with suppress(TimeoutError):
+                await wait_for(self.maintenance_needed.wait(), timeout=1)
+            self.maintenance_needed.clear()
+
+    async def add_one_connection(self):
+        try:
+            c = await Connection.create(self.origin, ssl=self.ssl)
+            self.active.add(c)
+            self.cert_hook(c)
+            return True
+        except Exception:
+            logger.exception("Failed creating APN connection")
```

Comment:
```suggestion
            logger.exception("Failed creating APN connection")
            return False
```

Reply:
None is Falsy, I don't see the point. 🤔

Reply:
just makes the return types the same which is neat. not required  of course